### PR TITLE
Gi

### PR DESCRIPTION
**Pull request by e2002e**
Reset my fork, was merged a while ago (probably when trying to merge GI into main).
Should be alles klär.

emissionVoxels

### CODE REVIEW

**Comment by eliecer696** [comment]
why don't they put this in armory yet

**Comment by e2002e** [comment]
Hi, I reworked the GI module so that forward emits light and render is the same than for deferred.

Screenshot from 2023-02-02 10-01-34
Screenshot from 2023-02-02 10-01-46

These two picture are showing respectively forward and deferred renderer.
Appart from that I did also make change so that SSRS and MicroShadowing are taken into account when rendering with forward renderer.

**Comment by e2002e** [comment]
Screenshot from 2023-02-03 11-38-28
Here, emissive material are lighting sphere from downward.
With SSRS, MicroShadowing and bloom ON.

**Comment by eliecer696** [comment]
how con I use this am taller of waiting for merge

**Comment by e2002e** [comment]
you clone m'y armsdk fork ans do a gît submodule update --init --remote --recursive

Le 3 février 2023 13:58:34 GMT+01:00, eliecer696 ***@***.***> a écrit :

**Comment by e2002e** [comment]
@eliecer696  https://github.com/e2002e/armsdk/tree/main

**Comment by eliecer696** [comment]
> Just take the armory folder and replace the original one from armory3d/armsdk with it.

am looking in render path post processing and i can't find n option to enable it

**Comment by e2002e** [comment]
It's not a screen espace effet. You must enable it from the voxels panel.

Le 3 février 2023 17:54:46 GMT+01:00, eliecer696 ***@***.***> a écrit :

**Comment by eliecer696** [comment]
> It's not a screen espace effet. You must enable it from the voxels panel. Le 3 février 2023 17:54:46 GMT+01:00, eliecer696 ***@***.***> a écrit :
> […](#)
> > Just take the armory folder and replace the original one from armory3d/armsdk with it. am looking in render path post processing and i can't find n option to enable it -- Reply to this email directly or view it on GitHub: [#2770 (comment)](https://github.com/armory3d/armory/pull/2770#issuecomment-1416136639) You are receiving this because you authored the thread. Message ID: ***@***.***>

this is all am getting 
awaw
voxe

**Comment by e2002e** [comment]
@MoritzBrueckner @luboslenco
I am finished here, global illumination supports refraction !
Screenshot from 2023-02-03 21-25-10
Screenshot from 2023-02-03 21-25-25
These are picture from both render mechanism. I'm abandoning the refraction module, it's getting nowhere.
Auf viehdersehen.

**Comment by e2002e** [comment]
Another last issue here, the specular reflection (which is supposed to render true reflections in fact) doesn't work.

**Comment by e2002e** [comment]
Reflections
Screenshot from 2023-02-05 18-49-48
 are working, yet we  got 
Screenshot from 2023-02-05 18-50-06
none from the environment. I'll check that next week.
Forward and deferred renderer.

**Comment by e2002e** [comment]
Screenshot from 2023-02-05 20-05-15

**Comment by e2002e** [comment]
@lubos there is a mistake in the way you handle environment lighting.
This is fixed with this pull request.
The voxels allow emissive material, true reflection/refraction. (I guess as it was some years ago but now with environment light too).

Screenshot from 2023-02-05 20-30-32

**Comment by e2002e** [comment]
Screenshot from 2023-02-06 13-39-27

**Comment by eliecer696** [comment]
> > It's not a screen espace effet. You must enable it from the voxels panel. Le 3 février 2023 17:54:46 GMT+01:00, eliecer696 _**@**_.***> a écrit :
> > […](#)
> > > Just take the armory folder and replace the original one from armory3d/armsdk with it. am looking in render path post processing and i can't find n option to enable it -- Reply to this email directly or view it on GitHub: [#2770 (comment)](https://github.com/armory3d/armory/pull/2770#issuecomment-1416136639) You are receiving this because you authored the thread. Message ID: _**@**_.***>
> 
> this is all am getting awaw voxe

armory is not loading 


File "C:\Users\mikewaski\Documents\armsdk - Copy/armory\blender\arm\material\mat_utils.py", line 51
    elif is_transluc(material) and not material.arm_discard and rpdat.rp_translucency_state != 'Off' and not material.arm_blending:
TabError: inconsistent use of tabs and spaces in indentation
  File "C:\Users\mikewaski\Documents\armsdk - Copy/armory\blender\arm\material\mat_utils.py", line 51
    elif is_transluc(material) and not material.arm_discard and rpdat.rp_translucency_state != 'Off' and not material.arm_blending:
TabError: inconsistent use of tabs and spaces in indentation

**Comment by e2002e** [comment]
Hi @eliecer696, you must use git clone https://github.com/e2002e/armory3d and put the folder in place of the original on inside the armsdk folder.

**Comment by eliecer696** [comment]
> Hi @eliecer696, you must use git clone https://github.com/e2002e/armory3d and put the folder in place of the original on inside the armsdk folder.
i did it man cloned the replaced it with the armory folder in the original sdk v 2023.2 and i get the error in terminal and all the armory don't initializes

**Comment by luboslenco** [comment]
Looks nice, but apparently some of the fixes done at https://github.com/armory3d/armory/pull/2676 are now missing here.

File: Shaders/deferred_light/deferred_light.json [hunk 1]
```diff
@@ -230,12 +230,6 @@
 					"link": "_biasLightWorldViewProjectionMatrixSpot2",
 					"ifndef": ["_ShadowMapAtlas"],
 					"ifdef": ["_LTC", "_ShadowMap"]
-				},
-				{
-					"name": "LWVPSpot[3]",
-					"link": "_biasLightWorldViewProjectionMatrixSpot3",
-					"ifndef": ["_ShadowMapAtlas"],
-					"ifdef": ["_LTC", "_ShadowMap"]
```
Comment (luboslenco) [inline Shaders/deferred_light/deferred_light.json:238]: Should this be removed?

File: Sources/armory/renderpath/Inc.hx [hunk 2]
```diff
@@ -224,8 +223,9 @@ class Inc {
 		#if arm_debug
 		endShadowsLogicProfile();
 		#end
-		#end // rp_shadowmap
+		#end //rp_shadowmap
```
Comment (luboslenco) [inline Sources/armory/renderpath/Inc.hx:226]: Please try to minimize the formatting changes. If there are formatting errors then a separate PR is still welcome.

File: Sources/armory/renderpath/RenderPathDeferred.hx [hunk 3]
```diff
@@ -560,6 +588,7 @@ class RenderPathDeferred {
 		path.bindTarget("_main", "gbufferD");
 		path.bindTarget("gbuffer0", "gbuffer0");
 		path.bindTarget("gbuffer1", "gbuffer1");
+		path.bindTarget("gbuffer_refraction", "gbuffer_refraction");
```
Comment (luboslenco) [inline Sources/armory/renderpath/RenderPathDeferred.hx:591]: Hm, gbuffer_refraction should be bound only when it is enabled?

File: blender/arm/make_renderpath.py [hunk 4]
```diff
@@ -303,7 +330,6 @@ def build():
                 wrd.compo_defs += '_BloomAntiFlicker'
 
         if rpdat.rp_ssr:
-            wrd.world_defs += '_SSR'
```
Comment (luboslenco) [inline blender/arm/make_renderpath.py:306]: Why remove this?

File: blender/arm/material/cycles.py [hunk 5]
```diff
@@ -131,7 +131,7 @@ def parse_material_output(node: bpy.types.Node, custom_particle_node: bpy.types.
                     arm.assets.add_khafile_def('rp_gbuffer_emission')
 
         if parse_opacity:
-            curshader.write('opacity = {0} - 0.0002;'.format(out_opacity))
+            curshader.write('opacity = {0};'.format(out_opacity))
```
Comment (luboslenco) [inline blender/arm/material/cycles.py:134]: Probably better to investigate this in a separate PR.

File: Sources/armory/renderpath/RenderPathDeferred.hx (hunk 3 above)
Comment (e2002e) [reply Sources/armory/renderpath/RenderPathDeferred.hx:591]: fine.

File: Shaders/deferred_light/deferred_light.json (hunk 1 above)
Comment (e2002e) [reply Shaders/deferred_light/deferred_light.json:238]: ok I set it back

File: blender/arm/material/cycles.py (hunk 5 above)
Comment (e2002e) [reply blender/arm/material/cycles.py:134]: Ok

File: Sources/armory/renderpath/Inc.hx [hunk 6]
```diff
@@ -444,15 +444,19 @@ class Inc {
 
 	#if rp_voxels
 	public static function initGI(tname = "voxels") {
-		#if arm_config
-		var config = armory.data.Config.raw;
-		if (config.rp_gi != true || voxelsCreated) return;
-		voxelsCreated = true;
-		#end
-
```
[the remark this answers is not in the document]
Comment (e2002e) [reply Sources/armory/renderpath/Inc.hx:452]: Restored.

[the remark this answers is not in the document]
Comment (e2002e) [reply Sources/armory/renderpath/Inc.hx:452]: restored

**Comment by e2002e** [comment]
Screenshot from 2023-02-06 17-41-22
Screenshot from 2023-02-06 17-41-37
These are render with high settings in forward and deferred renderer ; Sphere with various alpha/roughness options.

**Comment by e2002e** [comment]
Screenshot from 2023-02-06 18-38-09
Clustered lights !

**Comment by e2002e** [comment]
Screenshot from 2023-02-06 21-55-12
with reflection from the environment.

**Comment by e2002e** [comment]
There are a few last issues.
The first one is in the difference between the two renderers:
Screenshot from 2023-02-07 11-35-41
Screenshot from 2023-02-07 11-36-23

The second is with SSRS, which needs the depth. Since in forward render we have to use depth prepass. Can't get that working though.

When in deferred renderer, if there are transparent meshes and refraction isn't active and temporal filter is enabled, translucent meshes flicker.

The last thing to fix is with the refraction of the background.
I think this should be left apart from this pull request and be dealt with a screen space refraction pass.

**Comment by e2002e** [comment]
I tested screen space refraction along side voxels:
Screenshot from 2023-02-08 00-20-42

I will make a sepate pull request for this once I find a solution  to make it work with forward renderer.

**Comment by e2002e** [comment]
@eliecer696
You can use the module now !

**Comment by eliecer696** [comment]
> @eliecer696 You can use the module now !

congrats my friend its really good what you accomplished how you get the bounty 👌👍

**Comment by e2002e** [comment]
Temporal filter only works with deinterleaved buffer. I don't intend to fix this.

**Comment by e2002e** [comment]
Screenshot from 2023-02-12 06-37-21
Screenshot from 2023-02-12 06-37-32
Here are pictures of deferred and forward renderer with same settings, 
point light + sun (voxel does manage clusters).

**Comment by e2002e** [comment]
Hi, I've set the make_voxels.py to use same mechanism than forward materials to parse light values. The sampleLight function is therefore modified (telling us if used by voxel shaders or not). It works great you can test it with point lights and sun and will get the appropriate render.

Spot and area lights are not working.
Refraction does, there is also real reflection (objects out of the screen can show with enough details).

Bug with temporal filter if you use a transparent object without enabling refraction, and is in deferred renderer.

Le 6 février 2023 17:10:50 GMT+01:00, Lubos Lenco ***@***.***> a écrit :

**Review by e2002e — COMMENTED** [review]
It's good für mich

**Comment by e2002e** [comment]
Screenshot from 2023-02-12 18-47-05
Last thing I can do, use the alpha from coneTrace. Can't get it to work with forward renderer though.

**Comment by e2002e** [comment]
@eliecer696 stick to the GI branch, main has only screen space refraction more and one ; it's unstable, two there is already refraction with voxels now.

**Comment by eliecer696** [comment]
@e2002e am getting this error with the GI Branch 
 (node:11536) UnhandledPromiseRejectionWarning: Shader compiler error.
(node:11536) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:11536) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

Compiling shader 7 of 30 (armdefault_voxel.geom.glsl) failed:

**Comment by e2002e** [comment]
hi, try krom instead of nodejs.

Le 13 février 2023 02:15:26 GMT+01:00, eliecer696 ***@***.***> a écrit :

**Comment by eliecer696** [comment]
it works with nodejs krom is the one that gives me that error C:\Users\elio\Music\build_armory_playround\debug\temp\armdefault_voxel.geom.hlsl(27,9-14): error X3004: undeclared identifier 'output'

Compiling shader 6 of 29 (armdefault_voxel.geom.glsl) failed:
Shader compiler error.
(node:7172) UnhandledPromiseRejectionWarning: Shader compiler error.

**Comment by e2002e** [comment]
I don't get an arm_default shader.
Keep settings at default and enable voxelgi first, then add the others if that works.

**Comment by eliecer696** [comment]
i keep getting the same error

**Comment by e2002e** [comment]
@eliecer696 try will newly pushed code (I forgot something).

**Comment by eliecer696** [comment]
@e2002e error


C:\Users\me\Documents\armsdk-main_backup/nodejs/node.exe C:\Users\me\Documents\armsdk-main_backup/Kha/make krom -g direct3d11 --shaderversion 330 --parallelAssetConversion 4 --to build_armory_playround/debug --quiet
Info: For Armory to work correctly, you need Blender 3.3 LTS.

C:\Users\me\Music\build_armory_playround\debug\temp\armdefault_voxel.geom.hlsl(27,9-14): error X3004: undeclared identifier 'output'
C:\Users\elio\Music\build_armory_playround\debug\temp\armdefault_voxel.frag.hlsl(212,16-174): error X4000: variable 'n' used without having been completely initialized

Compiling shader 6 of 15 (armdefault_voxel.geom.glsl) failed:
Shader compiler error.
(node:9476) UnhandledPromiseRejectionWarning: Shader compiler error.
(node:9476) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:9476) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

Compiling shader 5 of 15 (armdefault_voxel.frag.glsl) failed:
Shader compiler error.
Finished in 1.817s

**Comment by e2002e** [comment]
I can't help you here. Try another driver (lastest) and/or switch to Ubuntu, got no problems here.

**Comment by e2002e** [comment]
Screenshot from 2023-02-17 00-59-01
Hi, just posting picture of voxel gi with two bounces. You can see the reflections on the table's legs.
Issue is that transparent materials do not render transparent shadows, so I had to deactivate "cast shadow" option for the light's glass area, windows don't show background either.

**Pull request closed without merging**